corehq/apps/sso/utils/entra.py

```diff
@@ -11,73 +11,131 @@ class MSGraphIssue:
-        response.raise_for_status()  # Raises an error for bad status
-        assignments = response.json()
-
-        # microsoft.graph.appRoleAssignment's property doesn't  userPrincipalName
-        # Property principalType can either be User, Group or ServicePrincipal
-        principal_ids = {assignment["principalId"] for assignment in assignments["value"]
-                         if assignment["principalType"] == "User"}
-
-        # Prepare batch request
-        batch_payload = {
-            "requests": [
-                {
-                    "id": str(i),
-                    "method": "GET",
-                    "url": f"/users/{principal_id}?$select=userPrincipalName"
-                } for i, principal_id in enumerate(principal_ids)
-            ]
-        }
-
-        # Send batch request
-        batch_response = requests.post(
-            'https://graph.microsoft.com/v1.0/$batch',
-            headers={'Authorization': 'Bearer ' + result['access_token'], 'Content-Type': 'application/json'},
-            data=json.dumps(batch_payload)
-        )
-        batch_response.raise_for_status()
-        batch_result = batch_response.json()
-
-        for resp in batch_result['responses']:
-            if 'body' in resp and 'error' in resp['body']:
-                raise EntraVerificationFailed(resp['body']['error']['code'], resp['body']['message'])
-
-        # Extract userPrincipalName from batch response
-        user_principal_names = [
-            resp['body']['userPrincipalName'] for resp in batch_result['responses']
-            if 'body' in resp and 'userPrincipalName' in resp['body']
-        ]
-        return user_principal_names
-    else:
+    if "access_token" not in result:
         raise EntraVerificationFailed(result.get('error', {}),
                                       result.get('error_description', 'No error description provided'))
+    return result.get("access_token")
+
+
+def get_all_user_ids_in_app(token, app_id):
+    endpoint = (f"{ENDPOINT_BASE_URL}/servicePrincipals(appId='{app_id}')/"
+               f"appRoleAssignedTo?$select=principalId, principalType")
+    # Calling graph using the access token
+    response = requests.get(
+        endpoint,
+        headers={'Authorization': 'Bearer ' + token},
+    )
+    response.raise_for_status()  # Raises an error for bad status
+    assignments = response.json()
+
+    user_ids = set()
+    group_queue = []
+
+    # Property principalType can either be User, Group or ServicePrincipal
+    for assignment in assignments["value"]:
+        if assignment["principalType"] == "User":
+            user_ids.add(assignment["principalId"])
+        elif assignment["principalType"] == "Group":
+            group_queue.append(assignment["principalId"])
+        else:
+            # Service Principal represents an application
+            # Which make the query too complicated
+            raise NotImplementedError("The application have Service Principal member")
```

Comment:
how will the user see this error message if this is raised? Is this an error message the user will see directly? Also, should one `ServicePrincipal` member trigger a failure for the whole task?

Suggested rewrite of error message:
```suggestion
            raise NotImplementedError("ServicePrincipal members are not supported. Please include only Users or Groups as members of this SSO application.")
```

Reply:
Hi Biyeun, user won't see this error message, this will only be raised as `MSGraphIssue.OTHER_ERROR`, should we send this to user? 
> should one ServicePrincipal member trigger a failure for the whole task?
Are you saying we still processed other members in this idp but skip the Service Principal only?

Reply:
```
Nested applications (ServicePrincipal members) are not supported. Please include only Users or Groups as members of this SSO application
```
We should send this to the user and we should not complete the task in this case, as the list of users would be incomplete

Reply:
Changed in 451b52407b2cd68a5a6a619ce6839731d59f889c